### PR TITLE
test: Increase code coverage in package jsapi-utils

### PR DESCRIPTION
**Pull request by emilyhuxng**
Increase code coverage in package jsapi-utils

### CODE REVIEW

File: __mocks__/dh-core.js [hunk 1]
```diff
@@ -466,17 +466,22 @@ class Format {
 }
 
 class FilterCondition {
+  constructor(type, parent) {
```
Comment (mofojed) [inline __mocks__/dh-core.js:469]: Doesn't look like you ended up using these, you just defined a new mock in `TableUtils.test.ts`. Revert the changes to this file.

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 2]
```diff
@@ -33,66 +36,78 @@ function makeColumns(count = 5): Column[] {
...
-  tableSorts = TableUtils.toggleSortForColumn(tableSorts, columns, 3, true);
-  table.applySort(tableSorts);
-
-  expect(table.sort.length).toBe(0);
-});
+type MockFilterCondition = Record<'not' | 'and' | 'or', jest.Mock>;
+function makeFilterCondition(type = ''): MockFilterCondition {
+  return {
+    not: jest.fn(
+      () => makeFilterCondition(`${type}.${FilterType.eq}`) as FilterCondition
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:43]: Shouldn't need the `as FilterCondition` for any of these

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 3]
```diff
@@ -175,6 +190,30 @@ describe('quick filter tests', () => {
...
+
+    filter[expectedFn].mockReturnValueOnce(expectedResult);
+
+    const result = TableUtils[functionName](column, text);
+
+    if (args.length > 0) {
+      expect(filter[expectedFn]).toHaveBeenCalledWith(...args);
+    } else {
+      expect(filter[expectedFn]).toHaveBeenCalled();
+    }
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:213]: Shouldn't need to split this case up. `...args` will just be an empty array, and calling `.toHaveBeenCalledWith(...args)` just means it will have been called with no arguments, which is actually a stricter check than just `.toHaveBeenCalled()` (which would pass whether it was called with or without arguments).

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 4]
```diff
@@ -243,7 +282,7 @@ describe('quick filter tests', () => {
     };
 
     dh.FilterCondition = {
-      invoke: jest.fn(type => makeFilterCondition(type)),
+      invoke: jest.fn(type => makeFilterCondition(type) as FilterCondition),
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:285]: Shouldn't need `as FilterCondition` here

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 5]
```diff
@@ -267,6 +306,381 @@ describe('quick filter tests', () => {
...
+
+      expect(filter.isNull).toHaveBeenCalled();
+      expect(nullResult.not).toHaveBeenCalled();
+      expect(notResult.and).toHaveBeenCalledWith(invokeResult);
+
+      if (args.length > 0) {
+        expect(filter.invoke).toHaveBeenCalledWith(...args);
+      } else {
+        expect(filter.invoke).toHaveBeenCalled();
+      }
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:442]: Shouldn't need to split this out... we should be strict about what it's getting called with.

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 6]
```diff
@@ -267,6 +306,381 @@ describe('quick filter tests', () => {
...
+        operation,
+        value,
+        timezone
+      );
+
+      if (args.length > 0) {
+        expect(filter[expectedFn]).toHaveBeenCalledWith(...args);
+      } else {
+        expect(filter[expectedFn]).toHaveBeenCalled();
+      }
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:473]: Don't split it out, let's be explicit about what the expected function gets called with.

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 7]
```diff
@@ -267,6 +306,381 @@ describe('quick filter tests', () => {
...
+        const invokeResult = makeFilterCondition();
+        const expectedResult = makeFilterCondition();
+
+        filter.isNull.mockReturnValueOnce(nullResult);
+
+        filter.invoke.mockReturnValueOnce(invokeResult);
+
+        invokeResult.not.mockReturnValueOnce(notResult);
+
+        (nullResult.or as jest.Mock).mockReturnValueOnce(expectedResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:624]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 8]
```diff
@@ -279,7 +693,9 @@ describe('quick filter tests', () => {
     it('handles empty cases', () => {
       const column = makeFilterColumn();
 
+      // @ts-expect-error text is generally a string
       expect(TableUtils.makeQuickNumberFilter(column, null)).toBe(null);
+      // @ts-expect-error text is generally a string
       expect(TableUtils.makeQuickNumberFilter(column, undefined)).toBe(null);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:699]: Instead of writing these test cases for invalid input (which should be caught by TypeScript anyway), just take out the lines in `makeQuickNumberFilter` where we check `text == null` (since it can never happen).

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 9]
```diff
@@ -615,8 +1038,10 @@ describe('quick filter tests', () => {
     it('handles invalid cases', () => {
       const column = makeFilterColumn();
 
+      // @ts-expect-error text is generally a string
       expect(() => TableUtils.makeQuickDateFilter(column, null, '')).toThrow();
       expect(() =>
+        // @ts-expect-error text is generally a string
         TableUtils.makeQuickDateFilter(column, undefined, '')
       ).toThrow();
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1046]: Don't need cases where we pass in `null` or `undefined` as TS shouldn't allow that.

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 10]
```diff
@@ -1168,11 +1593,11 @@ describe('quick filter tests', () => {
 
       filter.isNull.mockReturnValueOnce(nullResult);
 
-      nullResult.not.mockReturnValueOnce(notResult);
+      (nullResult.not as jest.Mock).mockReturnValueOnce(notResult);
 
       filter.invoke.mockReturnValueOnce(invokeResult);
 
-      notResult.and.mockReturnValueOnce(expectedResult);
+      (notResult.and as jest.Mock).mockReturnValueOnce(expectedResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1600]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 11]
```diff
@@ -1192,6 +1617,7 @@ describe('quick filter tests', () => {
       const column = makeFilterColumn();
 
       expect(TableUtils.makeQuickTextFilter(column, null)).toBe(null);
+      // @ts-expect-error text is generally a string
       expect(TableUtils.makeQuickTextFilter(column, undefined)).toBe(null);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1621]: We should just remove the `null` case in `makeQuickTextFilter`, it never gets called with `null` (or `undefined`)

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 12]
```diff
@@ -1211,7 +1637,9 @@ describe('quick filter tests', () => {
       filter.isNull.mockReturnValueOnce(expectedNullFilter);
 
       const expectedNotFilter = makeFilterCondition();
-      expectedNullFilter.not.mockReturnValueOnce(expectedNotFilter);
+      (expectedNullFilter.not as jest.Mock).mockReturnValueOnce(
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1640]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 13]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+      expect(TableUtils.makeSelectValueFilter(column, [], false)).toBe(
+        expectedResult
+      );
+      expect(filter.eq).toHaveBeenCalledWith(expectedValue);
+      expect(filter.notEq).toHaveBeenCalledWith(expectedValue);
+      expect(eqResult.and).toHaveBeenCalledWith(notEqResult);
+    };
+
+    it('should return null if there are no selected values and invertSelection is true', () => {
+      const column = {} as Column;
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1817]: Should probably still create an actual column here and not cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 14]
```diff
@@ -147,7 +162,7 @@ describe('quick filter tests', () => {
   function makeFilterColumn(type = 'string'): Column {
     const filter = makeFilter();
     const column = IrisGridTestUtils.makeColumn('test placeholder', type, 13);
-    column.filter = jest.fn(() => filter);
+    column.filter = jest.fn(() => filter as FilterValue);
     return column;
   }
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:167]: Much like `MockFilterCondition` is defined above, define `MockColumn` here - that will allow you to get rid of all the casting you have elsewhere. E.g.
```
  type MockColumn = Omit<Column, 'filter'> & { filter(): MockFilter };
  function makeFilterColumn(type = 'string'): MockColumn {
    const filter = makeFilter();
    const column = IrisGridTestUtils.makeColumn('test placeholder', type, 13);
    column.filter = jest.fn(() => filter);
    return column as MockColumn;
  }
```

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 15]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+    const testNoSelectedItems = (type: string, expectedValue: unknown) => {
+      const column = makeFilterColumn(type);
+      const filter = column.filter();
+
+      const eqResult = makeFilterCondition();
+      const notEqResult = makeFilterCondition();
+      const expectedResult = makeFilterCondition();
+
+      (filter.eq as jest.Mock).mockReturnValueOnce(eqResult);
+      (filter.notEq as jest.Mock).mockReturnValueOnce(notEqResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1805]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 16]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+      const column = makeFilterColumn('java.lang.String');
+      const filter = column.filter();
+
+      const isNullResult = makeFilterCondition();
+      const notResult = makeFilterCondition();
+      const notInResult = makeFilterCondition();
+      const expectedResult = makeFilterCondition();
+
+      (filter.notIn as jest.Mock).mockReturnValueOnce(notInResult);
+      (filter.isNull as jest.Mock).mockReturnValueOnce(isNullResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1842]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 17]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+    it('handles non-empty selected values with null values and invertSelection is false', () => {
+      const column = makeFilterColumn('boolean');
+      const filter = column.filter();
+
+      const isNullResult = makeFilterCondition();
+      const inResult = makeFilterCondition();
+      const expectedResult = makeFilterCondition();
+
+      (filter.isNull as jest.Mock).mockReturnValueOnce(isNullResult);
+      (filter.in as jest.Mock).mockReturnValueOnce(inResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1861]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 18]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+    });
+
+    it('handles all null values as selected values and invertSelection true', () => {
+      const column = makeFilterColumn('boolean');
+      const filter = column.filter();
+
+      const isNullResult = makeFilterCondition();
+      const expectedResult = makeFilterCondition();
+
+      (filter.isNull as jest.Mock).mockReturnValueOnce(isNullResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1877]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 19]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+      ).toBe(expectedResult);
+    });
+
+    it('handles all null values as selected values and invertSelection false', () => {
+      const column = makeFilterColumn('boolean');
+      const filter = column.filter();
+
+      const isNullResult = makeFilterCondition();
+
+      (filter.isNull as jest.Mock).mockReturnValueOnce(isNullResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1890]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 20]
```diff
@@ -1354,6 +1791,134 @@ describe('quick filter tests', () => {
...
+      ).toBe(isNullResult);
+    });
+
+    it('handles non-empty selected values but no null values and invertSelection is true', () => {
+      const column = makeFilterColumn('int');
+      const filter = column.filter();
+
+      const expectedResult = makeFilterCondition();
+
+      (filter.notIn as jest.Mock).mockReturnValueOnce(expectedResult);
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:1902]: Unnecessary cast

File: packages/jsapi-utils/src/TableUtils.test.ts [hunk 21]
```diff
@@ -1535,3 +2100,592 @@ describe('Sorting', () => {
...
+    modelColumn: number,
+    direction: SortDirection,
+    isAbs: boolean,
+    addToExisting: boolean,
+    expectedValue
+  ) => {
+    expect(
+      TableUtils.sortColumn(
+        // @ts-expect-error test what happens when columns is null
+        sorts,
```
Comment (mofojed) [inline packages/jsapi-utils/src/TableUtils.test.ts:2496]: No need to test `null` case

**Review by mofojed — CHANGES_REQUESTED** [review]

**Review by mofojed — APPROVED** [review]
